### PR TITLE
Grandfather in mobile workers without web apps

### PR DESCRIPTION
**Pull request by esoergel**
## Product Description
Fix web apps access for mobile workers

## Technical Summary
The problem was the introduction of this line into the code that filters which apps are available to users:
```python
    def access_web_app(self, app_id):
        return self.access_web_apps or app_id in self.web_apps_list
```
The old behavior was that the `access_web_apps` permission was needed to access web apps (EXCEPT for mobile workers, who got access no matter their role/permissions).  The new behavior was a parameterized permission list, so if you didn't allow access to _all_ web apps, you could provide access to some.

The expectation with this change was that all existing roles that needed it would have `access_web_apps` enabled, and going forward, projects could disable that permission and grant only specific apps that their users needed access to.  However, this didn't take into account that (I'd argue incorrect) behavior that there might be mobile workers accessing web apps even without the explicit permission to do so.

This fix works by only applying this filter if the user is a web user (who have always needed the permission) or someone whose role does provide explicit permissions for web apps.

### Safety story
Local testing and discussion with other developers

### Rollback instructions
- [x] This PR can be reverted after deploy with no further considerations

### Labels & Review
- [x] Risk label is set correctly
- [x] The set of people pinged as reviewers is appropriate for the level of risk of the change

### CODE REVIEW

File: corehq/apps/users/models.py [hunk 1]
```diff
@@ -1596,7 +1596,7 @@ def can_access_any_web_apps(self, domain=None):
         if domain:
             try:
                 role = self.get_role(domain)
-                return role.permissions.web_apps_list
+                return bool(role.permissions.web_apps_list)
```
Comment (millerdev) [inline corehq/apps/users/models.py:1599]: Is this change required for the fix, or is it cleanup?

File: corehq/apps/cloudcare/views.py [hunk 2]
```diff
@@ -118,7 +118,9 @@ def get_web_apps_available_to_user(self, domain, user):
         ))
         apps = filter(None, apps)
         apps = filter(lambda app: app.get('cloudcare_enabled') or self.preview, apps)
-        apps = filter(lambda app: user.can_access_web_app(domain, app.get('copy_of', app.get('_id'))), apps)
+        # Backwards-compatibility - mobile users haven't historically required this permission
+        if user.is_web_user() or user.can_access_any_web_apps(domain):
```
Comment (millerdev) [inline corehq/apps/cloudcare/views.py:122]: Would it make sense to make this contingent on `WEB_APPS_PERMISSIONS_VIA_GROUPS`?

Comment (esoergel) [reply corehq/apps/cloudcare/views.py:122]: Can you elaborate?  That flag enables the UI for the `app_access.user_can_access_app` filter below, whereas I think this behavior we want for all projects.

File: corehq/apps/users/models.py (hunk 1 above)
Comment (esoergel) [reply corehq/apps/users/models.py:1599]: just cleanup - I was reading this function to see what it did and I noticed this - shouldn't have any real effect though

File: corehq/apps/cloudcare/views.py (hunk 2 above)
Comment (orangejenny) [reply corehq/apps/cloudcare/views.py:122]: no, it shouldn't check that flag

**Review by orangejenny — APPROVED** [review]

File: corehq/apps/cloudcare/views.py (hunk 2 above)
Comment (millerdev) [reply corehq/apps/cloudcare/views.py:122]: > Can you elaborate?

I'm suggesting something like

_If the flag is disabled then use the new permissions models, otherwise ignore them and do what it did before._

I think I'm understanding correctly but find it confusing that the flag, when enabled, triggers the _old_ permissions model (via groups) and when disabled triggers the new permissions model via _permissions_?

Comment (orangejenny) [reply corehq/apps/cloudcare/views.py:122]: I think your description is right. I'm going to do a PR to access the flag within this function before checking the old style permissions, let's discuss on that PR.

Comment (orangejenny) [reply corehq/apps/cloudcare/views.py:122]: https://github.com/dimagi/commcare-hq/pull/34538

Comment (orangejenny) [reply corehq/apps/cloudcare/views.py:122]: Re-reading your comment, one adjustment: 
> the new permissions model via permissions instead of groups

The flag does not affect the roles/permissions configuration, so I'd write that as "in addition to", not "instead of"
